File: solutions/ocaml/086-sort-list.ml
(*
  Problem 86: Sort List (LeetCode 148)
  Difficulty: Med
  Language: OCaml
*)
let rec sort_list = function
  | [] -> []
  | [x] -> [x]
  | xs ->
    let rec split l r = function
      | [] -> (l, r) | [x] -> (x :: l, r)
      | a :: b :: t -> split (a :: l) (b :: r) t in
    let (l, r) = split [] [] xs in
    let rec merge a b = match a, b with
      | [], b -> b | a, [] -> a
      | x :: xs, y :: ys ->
        if x <= y then x :: merge xs (y :: ys)
        else y :: merge (x :: xs) ys in
    merge (sort_list l) (sort_list r)

File: solutions/elixir/086-sort-list.ex
# Problem 86: Sort List (LeetCode 148)
# Difficulty: Med
# Language: Elixir
# 
defmodule SortList do
  def sort([]), do: []
  def sort([x]), do: [x]
  def sort(list) do
    {l, r} = Enum.split(list, div(length(list), 2))
    merge(sort(l), sort(r))
  end

  defp merge([], b), do: b
  defp merge(a, []), do: a
  defp merge([a | as_], [b | bs]) when a <= b, do: [a | merge(as_, [b | bs])]
  defp merge(a, [b | bs]), do: [b | merge(a, bs)]
end
